### PR TITLE
Fixing remote resgistry insecure issues [fix for #1181, #1190, #561, #1015]

### PR DESCRIPTION
**Pull request by rishi-anand**
## Description

Adding support for insecure flag for external registry in `zarf init`. Fixing `zarf package create` and `zarf package deploy` failures with external self signed cert registry as mentioned in https://github.com/defenseunicorns/zarf/issues/1190 and https://github.com/defenseunicorns/zarf/issues/1181

## Related Issue

Fixes # https://github.com/defenseunicorns/zarf/issues/1190
Fixes # https://github.com/defenseunicorns/zarf/issues/1181
Fixes # https://github.com/defenseunicorns/zarf/issues/561
Fixes # https://github.com/defenseunicorns/zarf/issues/1015

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist before merging

- [ ] Tests have been added/updated as necessary (add the `needs-tests` label)
- [ ] Documentation has been updated as necessary (add the `needs-docs` label)
- [ ] An ADR has been written as necessary (add the `needs-adr` label) [ [1](https://github.com/joelparkerhenderson/architecture-decision-record) [2](https://cognitect.com/blog/2011/11/15/documenting-architecture-decisions) [3](https://adr.github.io/) ]

### CODE REVIEW

**Comment by jeff-mccoy** [comment]
Thanks @rishi-anand for working this; we are trying to wrap up #1173 tonight to fix an issue with external PRs unable to run and will get to this review after that is merged in.

File: src/cmd/package.go
Comment (jeff-mccoy) [inline src/cmd/package.go:268]: Good catch, must have been a victim of either the config file refactor or make zarf a library refactor.

File: src/pkg/packager/create.go
Comment (jeff-mccoy) [inline src/pkg/packager/create.go:214]: @YrrepNoj @Racer159 I wonder if we should consider changing to an overall `insecure` flag in `p.cfg.CommonOpts` and the root cmd vs all these random `insecure` flags we have floating around for different parts of Zarf?

File: src/types/runtime.go
Comment (jeff-mccoy) [inline src/types/runtime.go:37]: make sure you run `make docs-and-schema` and commit any changes when modifying types. Our `types` pkg  is used to generate the Typescript defs, JSON Schema & docs.

File: src/pkg/packager/create.go
Comment (Racer159) [reply src/pkg/packager/create.go:214]: I would agree with that.  It would simplify the logic and create (pull) and init/deploy (push) are already separate commands

File: src/types/runtime.go
Comment (rishi-anand) [reply src/types/runtime.go:37]: Sure @jeff-mccoy. I will do that.

File: src/pkg/packager/create.go
Comment (rishi-anand) [reply src/pkg/packager/create.go:214]: @jeff-mccoy, @YrrepNoj, @Racer159 Sure.. will do these changes in this PR itself.

File: src/types/runtime.go
Comment (rishi-anand) [reply src/types/runtime.go]: Changes are done in https://github.com/defenseunicorns/zarf/pull/1191/commits/ba3281bba2b85d66b1d322cdb5c6c02edc449f8c

File: src/pkg/packager/create.go
Comment (rishi-anand) [reply src/pkg/packager/create.go]: Changes are done in https://github.com/defenseunicorns/zarf/pull/1191/commits/ba3281bba2b85d66b1d322cdb5c6c02edc449f8c

File: src/config/lang/english.go
Comment (jeff-mccoy) [inline src/config/lang/english.go:40]:
```suggestion
	RootCmdFlagInseure     = "Allow access to insecure registries and disable other recommended security enforcements. This flag should only be used if you have a specific reason and accept the reduced security posture."
```

File: src/pkg/packager/deploy.go
Comment (jeff-mccoy) [inline src/pkg/packager/deploy.go:163]: Good add.

File: src/types/packager.go
Comment (jeff-mccoy) [inline src/types/packager.go:9]: Is there a reason this needs to be included here? We kept it in `config.CommonOptions` intentionally as it was shared amongst all instance of an execution (in case we ever have more than one package operation per run).

File: src/types/runtime.go
Comment (jeff-mccoy) [inline src/types/runtime.go:10]: Suggest use similar wording to the flag suggestion.

**Review by jeff-mccoy — CHANGES_REQUESTED** [review]
left a few notes, getting pretty close, thanks!

File: src/types/packager.go
Comment (rishi-anand) [reply src/types/packager.go]: Agree. Changes are in https://github.com/defenseunicorns/zarf/pull/1191/commits/6b1930dc8ef0850dffb6a635844e49b7e6bd44dd

**Comment by rishi-anand** [comment]
@jeff-mccoy It seems test cases have some issues for failed runs.

Screenshot 2023-01-21 at 11 34 18 PM

**Comment by jeff-mccoy** [comment]
That might be an issue with the insecure flag movement, I'll have to dig into it a bit more, but also you'll want to update https://github.com/spectrocloud/zarf/blob/fix-1181-1190/src/test/e2e/30_config_file_test.go as well since the flags are changing.

**Comment by Racer159** [comment]
> 

The errors in the screenshot are actually expected errors (we test that some of our validations work as expected).  The only thing that I see wrong is that by moving the config flag the defaults in the config file test have changed.

It is looking for:

```
Allow insecure registry connections when pulling OCI images (default true)
```

 in `30_config_file_test.go:123` but the text changed to:
 
 ```
 Allow access to insecure registries and disable other recommended security enforcements. This flag should only be used if you have a specific reason and accept the reduced security posture.
 ```
 
 What should fix it fully would be editing `/src/test/zarf-config.toml` to match the configuration changes you made (moving `--insecure` to the root) and then changing the text in the `30_config_file_test.go` test to match it (should be something like this:)
 
```
Allow access to insecure registries and disable other recommended security enforcements. This flag should only be used if you have a specific reason and accept the reduced security posture. (default true)
 ```

**Review by Racer159 — CHANGES_REQUESTED** [review]
Just a recommendation to help with code reuse

File: src/internal/packager/images/push.go
Comment (YrrepNoj) [inline src/internal/packager/images/push.go:49]: Would it make sense to add the `i.Insecure` as a parameter to this function? I think there are other places that might benefit from having the roundTripper added to the options (such as `tools.go -> func zarfCraneCatalog`)

File: src/cmd/package.go
Comment (YrrepNoj) [inline src/cmd/package.go:267]: This PR made me realize that the `shasum` variable we're setting this flag to isn't actually getting used. whoops..

Since the shasum flag is directly related to doing `--insecure` deployments, could you update this flag to be set to `pkgConfig.DeployOpts.Shasum`

**Review by YrrepNoj — CHANGES_REQUESTED** [review]

**Comment by jeff-mccoy** [comment]
We don't have push rights to your branch to make the requested changes, so changing to draft for now. Please mark ready for review again after the changes are made. Thanks!

File: src/internal/packager/images/push.go
Comment (rishi-anand) [reply src/internal/packager/images/push.go:49]: `GetCraneAuthOption` should only provide auth. Insecure and round ripper can be different options as mentioned in [this](https://github.com/defenseunicorns/zarf/pull/1191#discussion_r1084374638) review by @Racer159

File: src/cmd/package.go
Comment (rishi-anand) [reply src/cmd/package.go]: fixed in https://github.com/defenseunicorns/zarf/pull/1191/commits/4fb69aa44a9bd5dd894bca0ddc47711a25ced614

File: src/internal/packager/images/push.go
Comment (rishi-anand) [reply src/internal/packager/images/push.go]: fixed in https://github.com/defenseunicorns/zarf/pull/1191/commits/4fb69aa44a9bd5dd894bca0ddc47711a25ced614

**Comment by rishi-anand** [comment]
@jeff-mccoy changes are done as mentioned by you, @Racer159 and @YrrepNoj. Please review. I have tried to run `make test-e2e` but seeing some failures may be those are expected ones.

File: src/internal/packager/images/push.go
Comment (Racer159) [reply src/internal/packager/images/push.go:49]: I agree with @rishi-anand in keeping those separate for now.

**Review by Racer159 — CHANGES_REQUESTED** [review]
@rishi-anand the code overall looks good to me (very close!), but there is still an issue with 30_config_file_test.go on line 99:

```
		"Required if deploying a remote package and --shasum is not provided (default true)",
```

should be

```
		"Required if deploying a remote package and --insecure is not provided (default \"deploy.shasum: 7606fe19\")",
```

image

 Also the following will need to happen to get this fully merged:

1. You will need to merge in/resolve any conflicts from `main`
2. You will need to sign your PR commits as described in our contributor guide (#3 in the developer workflow): https://github.com/defenseunicorns/zarf/blob/main/CONTRIBUTING.md#developer-workflow

**Comment by jeff-mccoy** [comment]
looks like there is merge conflict that must be resolved before we can trigger the pipeline run

Screenshot 2023-01-24 at 9 03 32 PM

**Comment by rishi-anand** [comment]
@jeff-mccoy, Unfortunately, I can't resolve it.
image

**Comment by jeff-mccoy** [comment]
Hmmm. I think we can do it if you allow us to make changes to your branch, right now we don't have access.  Otherwise the only way to resolve this will be for you to resolve the conflict locally and push it. 

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by rishi-anand** [comment]
@jeff-mccoy fixed conflicts

**Comment by jeff-mccoy** [comment]
Migrated to #1280. We were still seeing merge conflicts and commits were unsigned (blocked by CI, see [contrib docs](https://docs.zarf.dev/docs/developer-guide/contributor-guide#developer-workflow)). I believe the main issue is the org settings on spectrocloud not allowing us to fix things, but just moving to our repo to get this wrapped up.

**Pull request closed without merging**
